### ISSUE TITLE
Stickiness seems to be not working on v0.43.1

### ISSUE THREAD

**Issue by cusspvz** [rancher/rancher#2520]
We need stickiness on LB since `socket.io` doesn't support `sid` sharing between node instances.

@deniseschannon can you please confirm if LB stickiness is working?

`rancher-compose.yml`

``` yaml
https:
  load_balancer_config:
    lb_cookie_stickiness_policy:
      cookie: stk
      domain: omitted
  default_cert: omitted
http:
  load_balancer_config:
    lb_cookie_stickiness_policy:
      cookie: stk
      domain: omitted
```

**Comment by deniseschannon** [comment]
Can you provide the haproxy.cfg for each load balancer? I notice that you have 2 in your rancher-compose.yml.

 http://docs.rancher.com/rancher/faqs/troubleshooting/#load-balancer

Is this the same issue as this one? https://github.com/rancher/rancher/issues/2443 This is only being addressed in our upcoming release.

**Comment by cusspvz** [comment]
Could be the same as #2443
Cookie settings are present on `http`, but `haproxy` seems to ignore it.

`http`
```
global
        log 127.0.0.1 local0
        log 127.0.0.1 local1 notice
        maxconn 4096
        maxpipes 1024
        chroot /var/lib/haproxy
        user haproxy
        group haproxy
        daemon

defaults
        log     global
        mode    tcp
        option  tcplog
        option  dontlognull
        option  redispatch
        option forwardfor
        retries 3
        timeout connect 5000
        timeout client 50000
        timeout server 50000
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

frontend 527563c5-5dc6-4bfb-88e2-116a549e79a0_frontend
        bind 10.42.18.230:80
        mode http

        acl 1_host hdr(host) -i host.domain.com
        acl 1_host hdr(host) -i host.domain.com:80
        acl 1_path path_beg -i /rsrcs/
        use_backend 527563c5-5dc6-4bfb-88e2-116a549e79a0_1_backend if 1_host 1_p
ath
        acl 0_host hdr(host) -i host.domain.com
        acl 0_host hdr(host) -i host.domain.com:80
        use_backend 527563c5-5dc6-4bfb-88e2-116a549e79a0_0_backend if 0_host

backend 527563c5-5dc6-4bfb-88e2-116a549e79a0_1_backend
        mode http
        balance roundrobin
        timeout check 2000

        cookie stk insert domain domain.com
        server 976b498e-db07-4d77-b5d0-5482807675db 10.42.3.210:8080 check port
8080 inter 2000 rise 2 fall 3 cookie 976b498e-db07-4d77-b5d0-5482807675db
        server 04592063-b6e2-4ec3-a38a-1dad2bb8fea9 10.42.104.186:8080 check por
t 8080 inter 2000 rise 2 fall 3 cookie 04592063-b6e2-4ec3-a38a-1dad2bb8fea9
        server 05c5fbc0-609b-4bf7-b018-63efae3a8f0a 10.42.16.98:8080 check port
8080 inter 2000 rise 2 fall 3 cookie 05c5fbc0-609b-4bf7-b018-63efae3a8f0a
        server 79f306a0-022d-4e36-a547-cf5f1d6fd2d6 10.42.190.91:8080 check port
 8080 inter 2000 rise 2 fall 3 cookie 79f306a0-022d-4e36-a547-cf5f1d6fd2d6
        server fad19859-eebd-4917-81bb-1ca548c0c9a5 10.42.65.124:8080 check port
 8080 inter 2000 rise 2 fall 3 cookie fad19859-eebd-4917-81bb-1ca548c0c9a5

backend 527563c5-5dc6-4bfb-88e2-116a549e79a0_0_backend
        mode http
        balance roundrobin
        timeout check 2000

        cookie stk insert domain domain.com
        server b5f7e352-a375-4423-a560-33c73f76beff 10.42.234.172:8080 check por
t 8080 inter 2000 rise 2 fall 3 cookie b5f7e352-a375-4423-a560-33c73f76beff
```

`https`
```
global
        log 127.0.0.1 local0
        log 127.0.0.1 local1 notice
        maxconn 4096
        maxpipes 1024
        tune.ssl.default-dh-param 2048
        ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-
GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES
128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECD
HE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES
256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA
:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-S
HA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-D
ES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES1
28-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5
:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
        ssl-default-server-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES12
8-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-A
ES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:E
CDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-A
ES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-S
HA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256
-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA
-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AE
S128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!M
D5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
        chroot /var/lib/haproxy
        user haproxy
        group haproxy
        daemon

defaults
        log     global
        mode    tcp
        option  tcplog
        option  dontlognull
        option  redispatch
        option forwardfor
        retries 3
        timeout connect 5000
        timeout client 50000
        timeout server 50000
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

frontend 68987f8d-b339-44cf-bf7f-8e8c58fc0ce3_frontend
        bind 10.42.40.236:443 ssl crt /etc/haproxy/certs/  no-sslv3 no-tlsv10
        mode http

        acl 1_host hdr(host) -i host.domain.com
        acl 1_host hdr(host) -i host.domain.com:443
        acl 1_path path_beg -i /rsrcs
        use_backend 68987f8d-b339-44cf-bf7f-8e8c58fc0ce3_1_backend if 1_host 1_p
ath
        acl 0_host hdr(host) -i host.domain.com
        acl 0_host hdr(host) -i host.domain.com:443
        use_backend 68987f8d-b339-44cf-bf7f-8e8c58fc0ce3_0_backend if 0_host

backend 68987f8d-b339-44cf-bf7f-8e8c58fc0ce3_1_backend
        mode http
        balance roundrobin
        timeout check 2000

        server 6ad104bc-d986-4d84-9822-48f5d5aefbaf 10.42.3.210:8080 check port
8080 inter 2000 rise 2 fall 3
        server 164f287b-0da9-4f75-8604-f5f027f1aeea 10.42.104.186:8080 check por
t 8080 inter 2000 rise 2 fall 3
        server d1123028-fd09-4b4a-9b71-a8cc91542268 10.42.16.98:8080 check port
8080 inter 2000 rise 2 fall 3
        server 821a4d32-06f9-4786-9521-6d769b25f33a 10.42.190.91:8080 check port
 8080 inter 2000 rise 2 fall 3
        server 68266aaa-22ac-430c-97c3-3cc7cbd42152 10.42.65.124:8080 check port
 8080 inter 2000 rise 2 fall 3
        http-request add-header X-Forwarded-Proto https if { ssl_fc }

backend 68987f8d-b339-44cf-bf7f-8e8c58fc0ce3_0_backend
        mode http
        balance roundrobin
        timeout check 2000

        server f3f04d46-0aae-4c72-8a11-19b089ea6006 10.42.234.172:8080 check por
t 8080 inter 2000 rise 2 fall 3
        http-request add-header X-Forwarded-Proto https if { ssl_fc }
```

Let me try to raise the LB service from the ground up again.

**Comment by cusspvz** [comment]
reset of entire LB stack didn't worked.

I've also created a new LB just for PoC proposes and it didn't included cookie settings on `haproxy.cfg`

```
...
frontend a791a94d-a0c2-49f1-9483-d190841dd697_frontend                          
        bind 10.42.241.141:8080 ssl crt /etc/haproxy/certs/  no-sslv3 no-tlsv10 
        mode http                                                               
                                                                                
        default_backend a791a94d-a0c2-49f1-9483-d190841dd697_0_backend          
                                                                                
backend a791a94d-a0c2-49f1-9483-d190841dd697_0_backend                          
        mode http                                                               
        balance roundrobin                                                      
        timeout check 2000                                                      
                                                                                
        server ed0e1c89-bb89-40b8-b35c-f65eca90839a 10.42.234.172:8080 check por
t 8080 inter 2000 rise 2 fall 3                                                 
        http-request add-header X-Forwarded-Proto https if { ssl_fc } 
```

**Comment by cusspvz** [comment]
Versions:
<img width="602" alt="captura de ecra 2015-11-3 as 19 25 26" src="https://cloud.githubusercontent.com/assets/3604053/10919080/adbb744e-8260-11e5-885b-5c2fcce07466.png">

**Comment by deniseschannon** [comment]
Could you try with v0.44.0 to see it is reproducible? Not sure why stickiness wouldn't work if it's configured in the haproxy.cfg.

**Comment by cusspvz** [comment]
> Is this the same issue as this one? #2443 This is only being addressed in our upcoming release.

@deniseschannon the stickiness issue on SSL is solved on `0.44.0`, but the one I was facing over http (the reason why I've posted this) is another.

Cookie configurations should differ between backends, but Rancher is setting them up as a Global setting.

Brief example: imagine a scenario where you have a server that handles frontend resources and another for WebSockets and API loadbalanced by the same port, `haproxy.cfg` will result into a frontend and two backends.

If you configure Stickiness on LB, it will set on both backends the same cookie.

1. Client starts socket handshake
2. LB sets cookie for sticking up with a WebSocket server
3. Client asks also for a frontend asset
4. LB replaces cookie for sticking up with a frontend server
5. Client tries to complete socket handshake
6. LB doesn't know which server was and justly does round-robin to another WebSocket server which won't know which handshake he is talking about.

So the issue here is that Stickiness cookie settings should differ between **haproxy backends** and should be part also of **Advanced Routing**.

As it is not a minor UI/agent change, I would propose to enumerate cookies based on global by backend:
If cookie is set as `sticky`, it would be `sticky-68987f8d` on `68987f8d-b339-44cf-bf7f-8e8c58fc0ce3_0_backend`, `sticky-a791a94d` on `a791a94d-a0c2-49f1-9483-d190841dd697_0_backend` and so on...

**Comment by cusspvz** [comment]
cc @sangeethah

**Comment by deniseschannon** [comment]
This is actually an enhancement to our load balancer, which I've labeled accordingly.

**Comment by cusspvz** [comment]
@deniseschannon is there any chance to place on the table my proposal of using the cookie name as a cookie prefix until cattle has the hability to differ cookie settings by backend?

This workaround should also be set up with a on/off checkbox so we do not mess those who are using their session cookies for sticky.

**Comment by alena1108** [comment]
With refactored LB, cookie prefix is different for every backend https://github.com/rancher/rancher/issues/2179

**Comment by cusspvz** [comment]
@alena1108 Nice. For those who got stuck on this, I've worked around by placing ws feature on a different service.

**Comment by sangeethah** [comment]
Tested with build from master

Lb created with following cookie policy works as expected:

```
   lbcookie_policy = {"mode": "insert",
                       "cookie": "cookie-1",
                       "indirect": True,
                       "nocache": True,
                       "postonly": False
                       }
```

haproxy.conf indicating that the cookies are configured  for every backend.

```
root@32457089431f:/# cat /etc/haproxy/haproxy.cfg
global
    chroot /var/lib/haproxy
    daemon
    group haproxy
    log 127.0.0.1:8514 local0
    log 127.0.0.1:8514 local1 notice
    maxconn 4096
    maxpipes 1024
    ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-
SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-
GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-E
CDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-
SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE
-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA25
6:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES-C
BC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-S
HA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PS
K:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
    ssl-default-bind-options no-sslv3 no-tlsv10
    ssl-default-server-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GC
M-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES12
8-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE
-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES25
6-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:D
HE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA
256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES
-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128
-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!
PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA
    tune.ssl.default-dh-param 2048
    user haproxy

defaults
    errorfile 400 /etc/haproxy/errors/400.http
    errorfile 403 /etc/haproxy/errors/403.http
    errorfile 408 /etc/haproxy/errors/408.http
    errorfile 500 /etc/haproxy/errors/500.http
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http
    errorfile 504 /etc/haproxy/errors/504.http
    log global
    maxconn 4096
    mode tcp
    option forwardfor
    option http-server-close
    option redispatch
    option tcplog
    retries 3
    timeout client 50000
    timeout connect 5000
    timeout server 50000

listen default
bind *:42

frontend 19023
bind *:19023
mode http
default_backend 19023_

backend 19023_
acl forwarded_proto hdr_cnt(X-Forwarded-Proto) eq 0
acl forwarded_port hdr_cnt(X-Forwarded-Port) eq 0
    http-request add-header X-Forwarded-Port %[dst_port] if forwarded_port
    http-request add-header X-Forwarded-Proto https if { ssl_fc } forwarded_prot
o
    cookie cookie-1_19023_ insert indirect nocache
mode http
server 10.42.211.28 10.42.211.28:80  cookie 10.42.211.28
server 10.42.125.203 10.42.125.203:80  cookie 10.42.125.203
```

@cusspvz  , this fix will be available in the upcoming 1.2 release.
